service/controller/azuremachinepool/handler/nodepool/create_scale_workers.go

```diff
@@ -139,42 +122,148 @@ func (r *Resource) scaleUpWorkerVMSSTransition(ctx context.Context, obj interfac
 	}
 
 	// We didn't scale up the VMSS, ready to move to next step.
-	return CordonOldWorkers, nil
+	return WaitForWorkersToBecomeReady, nil
 }
 
-func (r *Resource) scaleDownWorkerVMSSTransition(ctx context.Context, obj interface{}, currentState state.State) (state.State, error) {
-	azureMachinePool, err := key.ToAzureMachinePool(obj)
+func (r *Resource) splitInstancesByUpdatedStatus(ctx context.Context, azureMachinePool capzv1alpha3.AzureMachinePool) ([]compute.VirtualMachineScaleSetVM, []compute.VirtualMachineScaleSetVM, error) {
+	cluster, err := util.GetClusterFromMetadata(ctx, r.CtrlClient, azureMachinePool.ObjectMeta)
 	if err != nil {
-		return DeploymentUninitialized, microerror.Mask(err)
+		return nil, nil, microerror.Mask(err)
 	}
 
-	machinePool, err := r.getOwnerMachinePool(ctx, azureMachinePool.ObjectMeta)
-	if err != nil {
-		return DeploymentUninitialized, microerror.Mask(err)
+	if !cluster.GetDeletionTimestamp().IsZero() {
+		r.Logger.Debugf(ctx, "Cluster is being deleted, skipping reconciling node pool")
+		return nil, nil, nil
 	}
 
+	// All workers ready, we can scale up if needed.
+	var allWorkerInstances []compute.VirtualMachineScaleSetVM
+	{
+		r.Logger.Debugf(ctx, "finding all worker VMSS instances")
+
+		allWorkerInstances, err = r.GetVMSSInstances(ctx, azureMachinePool)
+		if err != nil {
+			return nil, nil, microerror.Mask(err)
+		}
+
+		r.Logger.Debugf(ctx, "found %d worker VMSS instances", len(allWorkerInstances))
+	}
+
+	resourceGroup := key.ClusterID(&azureMachinePool)
+	vmssName := key.NodePoolVMSSName(&azureMachinePool)
+
 	virtualMachineScaleSetsClient, err := r.ClientFactory.GetVirtualMachineScaleSetsClient(ctx, azureMachinePool.ObjectMeta)
 	if err != nil {
-		return currentState, microerror.Mask(err)
+		return nil, nil, microerror.Mask(err)
 	}
 
-	// Scale down to the desired number of nodes in worker VMSS.
-	desiredWorkerCount := *machinePool.Spec.Replicas
-	r.Logger.Debugf(ctx, "scaling worker VMSS to %d nodes", desiredWorkerCount)
+	vmss, err := virtualMachineScaleSetsClient.Get(ctx, resourceGroup, vmssName)
+	if err != nil {
+		return nil, nil, microerror.Mask(err)
+	}
 
-	strategy := scalestrategy.Quick{}
-	err = r.ScaleVMSS(ctx, virtualMachineScaleSetsClient, key.ClusterID(&azureMachinePool), key.NodePoolVMSSName(&azureMachinePool), int64(desiredWorkerCount), strategy)
+	var oldInstances []compute.VirtualMachineScaleSetVM
+	var newInstances []compute.VirtualMachineScaleSetVM
+	{
+		for _, i := range allWorkerInstances {
+			old, err := r.isWorkerInstanceFromPreviousRelease(ctx, cluster, azureMachinePool.Name, i)
```

Comment:
I think this part is still the same as earlier: Listing WC nodes for each VMSS instance.

It could be a good idea though to go with this implementation now and then refactor that separately so the change is clearer in diff.

Reply:
it's slightly changed but the main point is similar, yes.
I agree with a separate refactoring